File: leetcode/169_MajorityElement.cpp
#include <iostream>
#include <map>
using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        map<int,int> m;
        for (auto num : nums) {
            if (m.find(num) == m.end()) {
                m[num] = 1;
            } else {
                m[num]++;
            }
        }
        int num = 0;
        for (auto n : m) {
            if (n.second > nums.size()/2) {
                num = n.first;
                break;
            }
        }
        return num;
    }
};

int main(int argc, char* argv[]) {

    vector<int> nums = {2,2,1,1,1,2,2};

    Solution s;
    cout << s.majorityElement(nums) << endl;

    return 0;
}

File: leetcode/169_MajorityElement.rs
struct Solution;

impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {

        let mut count = 0;
        let mut candidate = 0;

        for num in nums {
            if count == 0 {
                candidate = num;
            }

            if num == candidate {
                count += 1;
            } else {
                count -= 1;
            }
        }

        candidate
    }
}

fn main() {

    let nums = [2,2,1,1,23,2];

    println!("{}", Solution::majority_element(nums.to_vec()));
}
